### PR TITLE
PHP 7 compatibility

### PR DESCRIPTION
**Pull request by kinglozzer**
Proposed solution to https://github.com/silverstripe/silverstripe-framework/issues/6759

### CODE REVIEW

**Comment by kinglozzer** [comment]
Okay, I’ve gone far enough down the rabbit hole for this evening. Outstanding issues are:

- segfault on 7.1 at ~25% test completion. I don’t know which test yet
- version constraints on reports module prevents CMS builds running

**Comment by sminnee** [comment]
* Pin the 7.1 tests to 7.1.2. We've had some issues with 7.1.3 on master.
 * Let's relax the version constraints for the supporting modules and re-run this.

File: model/fieldtypes/compat/Float.php [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?php
+
```
Comment (tractorcow) [inline model/fieldtypes/compat/Float.php]: Should we put a conditional on PHP_MAJOR_VERSION in here too? I'd move it to the start of this code from the DBFLoat.php file.


```
<?php

if (PHP_MAJOR_VERSION < 7) {
	return;
}

class Float extends DBFloat {
}
```

File: view/ViewableData.php [hunk 2]
```diff
@@ -331,7 +331,7 @@ public function castingClass($field) {
 	public function escapeTypeForField($field) {
 		$class = $this->castingClass($field) ?: $this->config()->default_cast;
 
-		return Config::inst()->get($class, 'escape_type', Config::FIRST_SET);
+		return Injector::inst()->get($class, true)->config()->escape_type;
```
Comment (tractorcow) [inline view/ViewableData.php]: Could we just use singleton()?

File: view/SSViewer.php [hunk 3]
```diff
@@ -536,7 +536,7 @@ public function getInjectedValue($property, $params, $cast = true) {
 					// If not provided, use default
 					if (!$casting) $casting = Config::inst()->get('ViewableData', 'default_cast', Config::FIRST_SET);
 
-					$obj = new $casting($property);
+					$obj = Injector::inst()->get($casting, false, array($property));
```
Comment (tractorcow) [inline view/SSViewer.php]: Shall we just use ->create() here? We normally use ->get() for singletons (just for consistency of style).

File: .travis.yml [hunk 4]
```diff
@@ -18,10 +18,12 @@ matrix:
     - php: 5.6
       env: DB=MYSQL PDO=1
     - php: 5.6
-      env: DB=MYSQL BEHAT_TEST=1
-    - php: 5.6
+      env: DB=MYSQL BEHAT_TEST=1 CMS_TEST=1
```
Comment (tractorcow) [inline .travis.yml]: Did you mean to remove the framework behat tests here? You're running cms behat tests twice now.

File: model/fieldtypes/compat/Float.php (hunk 1 above)
Comment (sminnee) [inline model/fieldtypes/compat/Float.php]: I don't think this will do anything as if the file is loaded you've already lost — with a compile-time error.

File: view/ViewableData.php (hunk 2 above)
Comment (sminnee) [inline view/ViewableData.php]: I prefer this style myself.

File: view/SSViewer.php (hunk 3 above)
Comment (sminnee) [inline view/SSViewer.php]: Agreed.

**Comment by sminnee** [comment]
^ PR for reports above.

**Comment by kinglozzer** [comment]
I also need to update the PHP version constraint to `<7.2`

File: model/fieldtypes/compat/Float.php (hunk 1 above)
Comment (tractorcow) [inline model/fieldtypes/compat/Float.php]: I'd rather a `class_exists` return false in PHP7 than an error be emitted.

File: view/ViewableData.php (hunk 2 above)
Comment (tractorcow) [inline view/ViewableData.php]: No problem then. :)

**Comment by sminnee** [comment]
@kinglozzer pushed a few commits

**Comment by sminnee** [comment]
One thing I'd note with this PR is that we now have 9 build-jobs, which feels a bit bloatly.

 * Shall we drop the PHP 5.4 and PHP 5.5 and SQLite builds? Keeping 5.3 and 5.6 should pick up most issues that affect 5.4 or 5.5.

File: model/fieldtypes/DBFloat.php [hunk 5]
```diff
@@ -60,3 +60,7 @@ public function prepValueForDB($value) {
 	}
 
 }
+
+if (PHP_MAJOR_VERSION < 7) {
+	include_once 'compat/Float.php';
```
Comment (tractorcow) [inline model/fieldtypes/DBFloat.php]: I still don't like classes being manually included in this way. It's not the job of DBFloat to trigger loading of Float.php.

File: model/fieldtypes/compat/Float.php (hunk 1 above)
Comment (tractorcow) [inline model/fieldtypes/compat/Float.php]: See my comment above. My major dislike is the roles that the php-7 equivalent fields in trying to load these deprecated classes.

File: model/fieldtypes/DBFloat.php (hunk 5 above)
Comment (sminnee) [inline model/fieldtypes/DBFloat.php]: I like this better than a manifest hack tbh

Comment (tractorcow) [inline model/fieldtypes/DBFloat.php]: I don't think we need to hack manifest.

**Comment by sminnee** [comment]
On the segfault, I believe that this has something to do with code in the LRU cache for Config. I had to patch this when getting PHP7 support working in SS4

File: model/fieldtypes/DBFloat.php (hunk 5 above)
Comment (sminnee) [inline model/fieldtypes/DBFloat.php]: What's your solution? Note that putting the check into the file won't block the parse error that PHP7 will throw.

Comment (tractorcow) [inline model/fieldtypes/DBFloat.php]: Oh, I hadn't realised. I thought PHP7 would safely escape at the "return". Can someone confirm this?

Comment (sminnee) [inline model/fieldtypes/DBFloat.php]: Confirmed with this test script:

```php
<?php

if(PHP_MAJOR_VERSION >= 7) {
	return;
}

if(PHP_MAJOR_VERSION < 7) {
	class Int {
	  function go() { echo "good"; }
	}
}

// $i = new Int();
// $i->go();
```

Results in

```
Fatal error: Cannot use 'Int' as class name as it is reserved in /Users/sminnee/Sites/ss3/framework/test-int.php on line 8
```

Comment (sminnee) [inline model/fieldtypes/DBFloat.php]: One issue with Loz' patch is that calling class_exists('Int') or a before class_exists('DBInt') or similar will return false, as the manifest won't pick up Int. Ditto subclassing Int, etc.

If that's a showstopper, one solution would be to load an additional autoloader if PHP < 7, which would load the Int and Float classes as needed.

I'd prefer an additional autoloader to hacking manifest, simply because it reduces the impact on pre-existing code.

**Comment by sminnee** [comment]
FYI I think this is the segfault fix: https://github.com/silverstripe/silverstripe-framework/pull/4551/commits/f5af0c85bade63174d282162a6c99fb37461b433

**Comment by sminnee** [comment]
We'll also need https://github.com/silverstripe/silverstripe-framework/pull/6232/commits/c80417a949c7f2821ab4ce1f76ccbc5b6649fcaf to fix the GridField test, but first we need a quick semver check.

File: model/fieldtypes/DBFloat.php (hunk 5 above)
Comment (tractorcow) [inline model/fieldtypes/DBFloat.php]: With a hammer-proof helmet on I present this solution.


```
<?php

if (PHP_MAJOR_VERSION < 7) {
    eval('class Int extends DBInt {}');
}
```

Still requires inclusion somewhere, as class manifest won't pick it up.

File: model/fieldtypes/compat/Float.php (hunk 1 above)
Comment (sminnee) [inline model/fieldtypes/compat/Float.php]: @kinglozzer @dhensby what do you think of this approach? I've added a separate autoloader for the Int/Float classes so as not to pollute manifest: https://github.com/silverstripe/silverstripe-framework/pull/6771/commits/c3ed81b9df0e278b85143fff59836b41ce3cdbb4

**Comment by tractorcow** [comment]
I propose merge on green. :) I'm happy with the solution now.

File: model/fieldtypes/compat/autoload.php [hunk 6]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+if (PHP_MAJOR_VERSION < 7) {
+	spl_autoload_register('php5_compat_autoloader');
+}
+
+function php5_compat_autoloader($classname) {
+	$classMap = [
```
Comment (tractorcow) [inline model/fieldtypes/compat/autoload.php]: Can't use php5 short array syntax in php5.3.

**Comment by tractorcow** [comment]
GridFieldTest::testGetStateData() is failing. I have a note in master regarding this test:

```
// @todo - PHP 7.0.6 change requires __isset() to return true
// for each reference from left to right along an isset() invocation.
// See https://bugs.php.net/bug.php?id=62059
```

Failing on https://github.com/kinglozzer/sapphire/blob/c3ed81b9df0e278b85143fff59836b41ce3cdbb4/tests/forms/GridFieldTest.php#L130

**Comment by sminnee** [comment]
@kinglozzer @dhensby I've made a lot of pushes to this branch to carry this forward.

The main things I did:
 * copy a few fixes from the SS4 compat work needed to get the tests passing (ViewableData changes, and removed use of SplFixedArray)
 * add an extra autoloader for Int and Float so that no failures occur if these classes are referenced before DBInt / DBFloat is loaded. added a version-aware test for this stuff too
 * addressed a few of Damian's suggestions

Still waiting for Travis but *hopefully* it's green. Feel free to rebase/squash this as you see fit if you want to keep hacking on it while I'm asleep. ;-)

**Comment by kinglozzer** [comment]
Looks like the current failures (a few cases of `Undefined property: X`) could have been caused by the magic getters behaviour change, that does make me a little nervous about it...

File: model/fieldtypes/compat/autoload.php [hunk 7]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+if (PHP_MAJOR_VERSION < 7) {
+	spl_autoload_register('php5_compat_autoloader');
```
Comment (kinglozzer) [inline model/fieldtypes/compat/autoload.php]: 👍

File: model/fieldtypes/compat/Float.php (hunk 1 above)
Comment (dhensby) [inline model/fieldtypes/compat/Float.php]: Wrapping in a condition like that doesn't actually work. PHP 7 seems to parse the file and throw a "can't name class Float" error before even evaluating the condition.

What benefit does the new auto-loader add? Just means we don't have a conditional include of a stub? I suppose that makes sense.

**Comment by dhensby** [comment]
> One thing I'd note with this PR is that we now have 9 build-jobs, which feels a bit bloatly.

@sminnee we can consolidate these because we have about 4 5.6 builds.. we only need one and can move the behat tests and cms tests about...

**Comment by dhensby** [comment]
> We'll also need c80417a to fix the GridField test, but first we need a quick semver check.

I'm happy this isn't a breaking change... we're just stopping bad edge-cases, aren't we?

**Comment by kinglozzer** [comment]
Right, more commits added:

- https://github.com/silverstripe/silverstripe-framework/pull/6771/commits/314fd406809e14ccd904d1031058560e1d49a70b - fixes an issue raised by the changes to `ViewableData::__isset()` as mentioned above. Similar fixes need to be added to CMS
- https://github.com/silverstripe/silverstripe-framework/pull/6771/commits/2c1736ae08f49e5f9fa723d53f27e933e595482f - copied from the `silverstripe/silverstripe-assets` module, stops PHP 7 complaining about performing numeric operations on strings like `512M`
- https://github.com/silverstripe/silverstripe-framework/pull/6771/commits/788ca7f7b8516f3ca18ef2d56d5637578960a6a0 - Fixes PHP 7.1+ `non-numeric value was encountered` error: https://travis-ci.org/silverstripe/silverstripe-framework/jobs/218382148#L550. tdlr; a unit test expects that calling `Image::SetHeight('String')` will raise an exception, it never got that far on PHP 7.1 because of the above warnings. `intval()` is equivalent to what PHP does if the error is suppressed anyway.

File: model/fieldtypes/compat/Float.php (hunk 1 above)
Comment (sminnee) [inline model/fieldtypes/compat/Float.php]: It means that you won't get a fatal error if you refer to class int without first referring to class dbint. The original approach suffered from that weakness.

**Comment by sminnee** [comment]
@kinglozzer you wanna do a similar PR for CMS. We're close!

**Comment by sminnee** [comment]
I'm comfortable with all Loz' commits; shall we git rebase -i this into a mergable shape?

**Comment by sminnee** [comment]
I recommend that Dan/Loz polish this today and then @tractor merges in the NZ morning.
#excite!

**Comment by dhensby** [comment]
@sminnee I'm working on the CMS fixes at the moment - it may not be done in time, though :(

**Comment by sminnee** [comment]
Yeah, it's a bit surprising that `$this->CopyContentFromID` on VirtualPage throws a warning — that's a has_one relationship on the object so it should return a value without a warning.

**Comment by kinglozzer** [comment]
I think we might be able to remove the `7.1.2` travis constraint - the segfault turned out to be unrelated to `7.1.3` specifically. I’ll let you guys take over again while I sleep 😄

**Comment by sminnee** [comment]
Ok I've pushed the 7.1 dep back to "7.1" and squashed some of the the fixup commits. I've left the commits that cherry-pick from SS4 broken out separately, to assist with tracking. I also added a bit more text on the initial commit.

**Comment by muskie9** [comment]
I like it if no major breakages... will likely move clients to a 7 environment once this is all sorted.

**Comment by kinglozzer** [comment]
:tada:🍾 thanks for the help @sminnee @dhensby @tractorcow
